File: yggdrasil/examples/rpcFib/src/rpcFibSrv.py
from __future__ import print_function
import sys
import numpy as np
from yggdrasil.interface.YggInterface import YggRpcServer
from yggdrasil.tools import sleep


def fibServer(args):

    sleeptime = float(args[0])
    print('Hello from Python rpcFibSrv: sleeptime = %f' % sleeptime)

    # Create server-side rpc conneciton using model name
    rpc = YggRpcServer("rpcFibSrv", "%d", "%d %d")

    # Continue receiving requests until error occurs (the connection is closed
    # by all clients that have connected).
    while True:
        print('rpcFibSrv(P): receiving...')
        retval, rpc_in = rpc.rpcRecv()
        if not retval:
            print('rpcFibSrv(P): end of input')
            break

        # Compute fibonacci number
        print('rpcFibSrv(P): <- input %d' % rpc_in[0], end='')
        pprev = 0
        prev = 1
        result = 1
        fib_no = 1
        arg = rpc_in[0]
        while fib_no < arg:
            result = prev + pprev
            pprev = prev
            prev = result
            fib_no = fib_no + 1
        print(' ::: ->(%2d %2d)' % (arg, result))

        # Sleep and then send response back
        sleep(float(sleeptime))
        flag = rpc.rpcSend(arg, np.int32(result))
        if not flag:
            raise RuntimeError('rpcFibSrv(P): ERROR sending')

    print('Goodbye from Python rpcFibSrv')

    
if __name__ == '__main__':
    fibServer(sys.argv[1:])

File: yggdrasil/examples/rpcFib/src/rpcFibSrv.c

#include "YggInterface.h"
#include <stdio.h>


int main(int argc, char *argv[]) {

  float timeSleep = atof(argv[1]);
  printf("Hello from C rpcFibSrv: sleeptime = %f\n", timeSleep);
  
  // Create server-side rpc conneciton using model name
  yggRpc_t rpc = yggRpcServer("rpcFibSrv", "%d", "%d %d");

  // Continue receiving requests until error occurs (the connection is closed
  // by all clients that have connected).
  int input;
  while (1) {
    printf("rpcFibSrv(C): receiving...\n");
    int ret = rpcRecv(rpc, &input);
    if (ret < 0) {
      printf("rpcFibSrv(C): end of input\n");
      break;
    }

    // Compute fibonacci number
    printf("rpcFibSrv(C): <- input %d", input);
    int result = 1;
    int prevResult = 1;
    int prevPrev = 0;
    int idx = 1;
    while(idx++ < input){
      result = prevResult + prevPrev;
      prevPrev = prevResult;
      prevResult = result;
    }
    printf(" ::: ->(%2d %2d)\n", input, result);

    // Sleep and then send response back
    if (timeSleep) 
      sleep(timeSleep);
    int flag = rpcSend(rpc, input, result);
    if (flag < 0) {
      printf("rpcFibSrv(C): ERROR sending\n");
      break;
    }
  }

  printf("Goodbye from C rpcFibSrv\n");
  return 0;
}

